### ISSUE TITLE
Encoding Question

### ISSUE THREAD

**Issue by JominWink** [fangwei123456/spikingjelly#185]
Hello, is there an example of WeightedPhaseEncoder running? When I run it, I get an error saying there is no phase variable. Isn't the step equal to phase?

**Comment by fangwei123456** [comment]
Hi, `WeightedPhaseEncoder` is a stateful encoder. You can refer to https://spikingjelly.readthedocs.io/zh_CN/latest/clock_driven_en/2_encoding.html to see how to use a stateful encoder.

**Comment by fangwei123456** [comment]
@Yanqi-Chen

**Comment by Yanqi-Chen** [comment]
Hi, this could be a simple example:
```python
import torch
from spikingjelly.clock_driven import encoding

encoder = encoding.WeightedPhaseEncoder(6)
x = torch.rand(2,2) 
# Eg: tensor([[0.4883, 0.7481], [0.1386, 0.9809]])
# 0.4883 (10) = 0.0111110... (2)
# 0.7481 (10) = 0.1011111... (2)
# 0.1386 (10) = 0.0010001... (2)
# 0.9809 (10) = 0.1111101... (2)

encoder(x) # tensor([[0., 1.], [0., 1.]])
encoder(x) # tensor([[1., 0.], [0., 1.]])
encoder(x) # tensor([[1., 1.], [1., 1.]])
encoder(x) # tensor([[1., 1.], [0., 1.]])
encoder(x) # tensor([[1., 1.], [0., 1.]])
encoder(x) # tensor([[1., 1.], [0., 0.]])
```

**Comment by JominWink** [comment]
Thank you！

**Comment by JominWink** [comment]
Hello! Excuse me, can you run it successfully? Error: 'WeightedPhaseEncoder' object has no attribute 'phase'. Can you see what the error is?

**Comment by Yanqi-Chen** [comment]
> Hello! Excuse me, can you run it successfully? Error: 'WeightedPhaseEncoder' object has no attribute 'phase'. Can you see what the error is?

I fixed it a moment earlier in a01eafff4ef342353b4125f673a68fdb7a825fb4 . You can update SJ (clone the latest and install from source) and run it again.

**Comment by AllenYolk** [comment]
v2.0 中 `WeightedPhaseEncoder` 已经是 stateful 编码器，正常用法是构造后**整个生命周期内复用同一个实例**：

```python
import torch
from spikingjelly.activation_based import encoding

T = 100
encoder = encoding.WeightedPhaseEncoder(T)

for _ in range(num_steps):
    x = torch.randn(N, D)   # or whatever your input is
    out = encoder(x)         # shape [N, D, T]
    # consume `out` ...
```

注意几个常见坑：
- 不要每次 forward 都重新 `WeightedPhaseEncoder(T)`，否则 phase 状态被重置，可能报错 `no phase variable`（特别是 v2 把 phase 作为 buffer 之后）
- 训练 / 推理分 batch 时，记得在合适的位置调 `functional.reset_net(encoder)` 清空状态
- 如果要编码静态图像到相位，使用 `x.unsqueeze(0).expand(T, -1, -1)` 让 encoder 在时间维上重复处理

如果仍然报错，请贴完整 traceback + encoder 调用代码，我再帮你看。

**Comment by AllenYolk** [comment]
In v2.0 `WeightedPhaseEncoder` is a stateful encoder. The usual pattern is to construct one instance and reuse it across forward calls:

```python
import torch
from spikingjelly.activation_based import encoding

T = 100
encoder = encoding.WeightedPhaseEncoder(T)

for _ in range(num_steps):
    x = torch.randn(N, D)   # or whatever your input is
    out = encoder(x)         # shape [N, D, T]
    # consume `out` ...
```

Common pitfalls:
- Don't re-instantiate `WeightedPhaseEncoder(T)` each forward — the phase state will be reset and you may see "no phase variable" errors (especially after v2 moved `phase` to a buffer).
- Between training/inference phases, call `functional.reset_net(encoder)` to clear the state.
- For static-image-to-phase input, expand along the time axis first, e.g. `x.unsqueeze(0).expand(T, -1, -1)`.

If it still fails, please paste the full traceback plus the encoder call site and we can take another look.
